### ISSUE TITLE
64-bit fscAnyCpu.exe uses more stack than 32-bit fsc.exe (was: Project that builds in VS2019 fails to build in VS2022)

### ISSUE THREAD

**Issue by kentcb** [dotnet/fsharp#12322]
A project that is part of a reasonably complex solution is failing to build in 2022 due to the compiler crashing. I've managed to boil things down to a simple, standalone repro shown below.

**Repro steps**

The following simplified code typifies that which is causing the compiler to crash:

```fsharp
module repro_fscanycpu_failure

//
// This builds fine in VS2019, but fails under VS2022 with:
//
// 1>  Process is terminated due to StackOverflowException.
// 1>  C:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\IDE\CommonExtensions\Microsoft\FSharp\Tools\Microsoft.FSharp.Targets(317,9): error MSB6006: "fscAnyCpu.exe" exited with code -1073741571.
//
// Opening this file in the VS2022 editor will also crash, presumably for the same underlying reason.
//

type ReproBuilder () =
    member _.Delay x = x ()
    member _.Yield  (x) = x
    member _.Combine (x, y) = x + y

let repro = ReproBuilder ()

let callSite () =
    repro {
        // Commenting out some of the below is enough to avoid StackOverflow on my machine.
        0
        1
        2
        3
        4
        5
        6
        7
        8
        9
        0
        1
        2
        3
        4
        5
        6
        7
        8
        9
        0
        1
        2
        3
        4
        5
        6
        7
        8
        9
        0
        1
        2
        3
        4
        5
        6
        7
        8
        9
        0
        1
        2
        3
        4
    }
```

**Expected behavior**

It should build in 2022 as it does in 2019.

**Actual behavior**

It fails with:

> 1>C:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\IDE\CommonExtensions\Microsoft\FSharp\Tools\Microsoft.FSharp.Targets(317,9): error MSB6006: "fscAnyCpu.exe" exited with code -1073741571.

**Known workarounds**

1. **Short-term fix**: Use VS2019.

2. **Recommended**: Make your inputs smaller, probably half the size will do, divide your large expressions into two

**Related information**

Provide any related information (optional):

* Windows 10
* .NET Core

**Comment by smoothdeveloper** [comment]
@kentcb just in case it has an impact, can you try compiling the same project in Release mode (assuming it was Debug mode you tried)?

**Comment by kentcb** [comment]
@smoothdeveloper Alas, same problem.

**Comment by eriawan** [comment]
@kentcb 
I think this is something related to TFM of the nuget package you used.
Have you update the microsoft.servicefabric nuget package? I've used version 8.1.335 and I have no problem.

See also
https://www.nuget.org/packages/Microsoft.ServiceFabric/8.1.335

**Comment by cartermp** [comment]
@kentcb if you're allowed to build with the `/bl` flag and upload the binlog then that makes it a lot easier for folks to diagnose.

**Comment by vzarytovskii** [comment]
> @kentcb if you're allowed to build with the `/bl` flag and upload the binlog then that makes it a lot easier for folks to diagnose.

Yeah, binlog or minimal repro would be ideal.

**Comment by kentcb** [comment]
@eriawan Yes, I did try upgrading to latest `Microsoft.ServiceFabric.Services.Remoting` to no avail. I've also tried to extract this into a minimal repro, but no luck.

Happy to upload a binlog, but is there somewhere private I can upload it or someone I can share a OneDrive link with?

**Comment by kentcb** [comment]
By building with more diagnostics, I could see that `fsAnyCpu.exe` was terminating due to stack overflow when building my `Types` project. By process of elimination, I identified which file was causing the crash. Even opening this file in the editor would eventually crash VS, so I had to edit it in VS2019, but doing so allowed me to identify the code that was triggering the regression...

It's a custom CE we use to define codecs. We use it extensively throughout our entire solution, and yet this one particular case was causing a SO. By playing around with it, I determined that there was no particular line triggering the SO, but rather the sheer number of them. It was happy with up to 38 lines, but the 39th would trigger a crash (either in editor, or in compiler, so presumably the same SO crash).

Of course, there must be something our CE is doing, or something about what we're doing on each line to trigger the SO. That's where I'm at and will possibly handball this issue internally.

**Comment by jimfoye** [comment]
Perhaps  related to #9231?

**Comment by kentcb** [comment]
@jimfoye Hmm, _maybe_, but I use Windows and sounds like it's a non-issue there. I'll try and find some time to repro this later in the week.

**Comment by kentcb** [comment]
@vzarytovskii @cartermp Please see updated issue description for a simple repro

**Comment by dsyme** [comment]
@kentcb Thanks for the repro.  We'll work on it, we don't have a large-input stress test of that size, and we must be doing something quadratic or worse.

**Comment by ly29** [comment]
Noticed something similar which crashes Visual Studio 2022 hard.

**Comment by patroeper** [comment]
We are seeing something similar wrt a StackOverflowException in an F# project when switching from VS2019 to VS2022. I have a repro of this at https://github.com/patroeper/fs_fsc_soe [only using let bindings in a module](https://github.com/patroeper/fs_fsc_soe/blob/main/Library1/Library1.fs). Let me know if I should carve this off as a separate issue.

**Comment by brianrourkeboll** [comment]
I get the same error (fscAnyCpu.exe exiting with -1073741571) in VS 2022 that builds in VS 2019. It does seem to be computation-expression related&mdash;in my case, it seems to be the `gen` computation expression from FsCheck.

Everything does still seem to build fine on the command line with `dotnet build`.

I also experience VS crashing almost instantly when I open certain files containing usage of the `gen` computation expression&mdash;or if I open any other files that are downstream of them in compilation order, or if I do a rename on things that happen to be referenced in such files.

**Comment by jimfoye** [comment]
Yeah I am getting the -1073741571 error when building FSharp.Data in VS 2022.

**Comment by dsyme** [comment]
Thank you for these reports, we will get this fixed

**Comment by dsyme** [comment]
A likely workaround is to set these (only the second is strictly needed, the first is the default). I'd be grateful if you could try these:

```
    <FSharpPreferNetFrameworkTools>true</FSharpPreferNetFrameworkTools>
    <FSharpPrefer64BitTools>false</FSharpPrefer64BitTools>
```

For the first repro up above:
* F# 5 net472 VS2019 Release fsc.exe 32-bit blows stack with **> ~85** nested Combine/Yield in computation expression
* F# 6 net472 VS2022 Release fsc.exe 32-bit blows stack with **> ~120** nested Combine/Yield in computation expression
* F# 6 net472 VS2022 Release fscAnyCpu.exe 64-bit blows stack with **> ~50 nested** Combine/Yield in computation expression
* F# 5 net5 SDK fsc.dll 64-bit blows stack with **> ~65 nested** Combine/Yield in computation expression
* F# 6 net5 SDK fsc.dll 64-bit blows stack with **> ~60 nested** Combine/Yield in computation expression

So it seems that F# 6 code is not really at faultactually consumes less stack (since net472 now process larger inputs), but the switch to 64-bit compiler by default is causing 2x stack to be used.

Note that 

> I get the same error (fscAnyCpu.exe exiting with -1073741571) in VS 2022 for a solution that builds in VS 2019. It does seem to be computation-expression related—in my case, it seems to be the gen computation expression from FsCheck.

@brianrourkeboll Do you have a public repro?

> Yeah I am getting the -1073741571 error when building FSharp.Data in VS 2022.

Thanks, we will use this to form the basis of a repro too

**Comment by dsyme** [comment]
Note that this workaround:

```
    <FSharpPreferNetFrameworkTools>true</FSharpPreferNetFrameworkTools>
    <FSharpPrefer64BitTools>false</FSharpPrefer64BitTools>
```

is **not** sufficient to stop IDE failures, because the IDE is now running as 64-bit.  Instead you should simply make your problematic inputs smaller as a workaround.  

I suppose we must work on linearizing the processing of the above.

**Comment by dsyme** [comment]
@jimfoye I investigated the failure building Debug FSharp.Data - see https://github.com/fsprojects/FSharp.Data/pull/1411 for the details, and it's related to the separate problem #12413

**Comment by brianrourkeboll** [comment]
@dsyme I'm not sure if this is what you were looking for, or whether it adds anything beyond @kentcb's minimal repro above, but the following snippet with 115 `let!`s in [FsCheck's `gen` computation expression](https://github.com/fscheck/FsCheck/blob/dbe1630e8037140decbd91ed2e1367f51415c928/src/FsCheck/Gen.fs#L49-L103) will crash VS 2022 if you open it in an F# script file (which however compiles fine when passed into `dotnet fsi` directly). The ~114 `let!` limit for the `gen` computation expression accords with what I see in my actual projects (i.e., files with fewer `let!`s are fine, but files with more cause VS 2022 to crash).

<details>

<summary>Repro</summary>

```fsx
#r "nuget: FsCheck"

open FsCheck

let g =
    gen {
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        let! _ = Arb.generate<int>
        return ()
    }
```

</details>

If you remove one `let!`, leaving only 114, it will not crash VS.

---

Separately, adding `<FSharpPrefer64BitTools>false</FSharpPrefer64BitTools>` to the affected fsproj files as you suggested above does indeed allow them to build in VS (as long as you don't open the files with the big computation expressions).

---

It may also be worth noting that I have another (custom, unrelated) computation expression in a separate project which in one case is used with more than 200 `and!`s&mdash;this also does not compile in VS 2022, but it does not crash VS, and the failure to compile seems to be for a different reason:

```console
error FS0073: internal error: Expression is too large and/or complex to emit. Method name: '...'. Recursive depth: 140.
```

I at first suspected that the difference between this case and that of `gen` is that most of this computation expression builder's methods are marked `inline`, while those for `gen` are not. That appears not to be the case, though: it just seems that the compiler will tolerate more `and!`s (closer to ~300) than `let!`s.

In this case, however, setting `FSharpPreferNetFrameworkTools` to `false` allows my project to build without error in VS 2022 (it builds without error in VS 2019 and with `dotnet build` using the .NET 5 or 6 SDKs).

**Comment by dsyme** [comment]
@brianrourkeboll Thank you 

Realistically our three shipped compilers (netfx 32, netfx 64, net6) do have different capacity limits along some input dimensions.  Ideally the capacity limits would be identical across all three , but that's really, really hard to arrange without avoiding the use of stack altogether.

I will look at the specific case of these computation expressions.  We do have the option of simply switching to a new stack, for example, at a particular expression depth.

**Comment by dsyme** [comment]
This is fixed in #12420, at least for practical purposes, given the current repros.

**Comment by jayhill** [comment]
Wow — impressive @dsyme!

Would this also somehow mitigate [this situation and repro](https://github.com/dotnet/fsharp/issues/12322#issuecomment-965710784), in which only `let` bindings — not custom CEs — are in play?

That is the same issue that I'm experiencing.

**Comment by dsyme** [comment]
> Would this also somehow mitigate this situation and repro, in which only let bindings — not custom CEs — are in play?

No, but we can get that fixed too

**Comment by dsyme** [comment]
@jayhill @patroeper Thank you for the repro, it was a new problem related to writing debug scopes.  I'm not sure why our existing testing didn't catch it as we certainly have test with hundreds or thousands of `let` .  Hmmmm....

In any case, a fix and testing is included in #12420

**Comment by baronfel** [comment]
@dsyme should this be closed since #12420 was merged?
